# Smallest Multiple

## Problem Statement -
$2520$ is the smallest number that can be divided by each of the numbers from 
$1$ to $10$ without any remainder.

What is the smallest positive number that is **evenly divisible** by all of the numbers from 
**1** to **20**?

## Answers -

## Vini's-logic (human with search assistance, some hours)

imports:

In [162]:
import math

- First, i want to explore how the **2520** relates with the numbers $1\to 10$

In [163]:
for i in range (1, 11):
    print(f"2520/{i}: {2520/i}")

2520/1: 2520.0
2520/2: 1260.0
2520/3: 840.0
2520/4: 630.0
2520/5: 504.0
2520/6: 420.0
2520/7: 360.0
2520/8: 315.0
2520/9: 280.0
2520/10: 252.0


- It would be fun to make a big for loop and 10 module statements checkig if its the correct number
- The difference between the numbers above are:

In [164]:
for i in range (1, 10):
    print(f"{2520/i}-{2520/(i+1)}: {int(2520/i - 2520/(i+1))}")

2520.0-1260.0: 1260
1260.0-840.0: 420
840.0-630.0: 210
630.0-504.0: 126
504.0-420.0: 84
420.0-360.0: 60
360.0-315.0: 45
315.0-280.0: 35
280.0-252.0: 28


- This doesn't help me, there's some minor relations between them but nothing imporant ($28=7\cdot4
, 35=7\cdot5$ and $210\cdot2=420, 420\cdot3=1260$)
- Another cool relation, after the even $28$ we have three numbers that end up in $0$ or $5$, then two even $84$ and $126$ and again three with $0$ or $5$
- i have no other idea of a relation that could give us an answer.
- LET'S BRUTE FORCE IT!!!

- Below i made a simple function to check for divisibility to all number $1\to20$

In [165]:
def isDivisibleByAllOfThis(x):
    for i in range(1, 21):
        if x%i!=0:
            return False
    return True

- Then i'll call it in a main function

In [166]:
def main():
    number = 2
    while not isDivisibleByAllOfThis(number):
        number+=1
    return number


- Mega inefficient, I'll do a bit of research
- Let's try prime factorization (im looking at some sources now, not my thougts only)

In [167]:
def is_prime(x):
    for i in range(2, math.isqrt(x) + 1):
        if x%i==0:
            return False
    return True

In [168]:
prime_numbers = [x for x in range (2, 21) if is_prime(x)]
prime_numbers

[2, 3, 5, 7, 11, 13, 17, 19]

- now let's make a function to count prime factors

In [169]:
def prime_factors(x):
    factors = {prime:0 for prime in prime_numbers}
    for prime in prime_numbers:
        while x%prime==0:
            factors[prime]+=1;
            x //=prime
    return factors

- And now use it the new function

In [170]:
factors = {prime: 1 for prime in prime_numbers}

for i in range(2,21):
    for prime, count in prime_factors(i).items():
       factors[prime] = max(count, factors[prime]) 

result = 1
for prime, count in factors.items():
    result *= prime ** count

print(result)

232792560


- For this answer i studied this souce: https://www.grae.io/post/euler_problem_5/
- The code is very similar, but that's how I learn, practice.

## O3-mini (6 seconds)


### Finding the Smallest Number Divisible by 1 to 20

This notebook computes the smallest positive number that is evenly divisible by all of the numbers from 1 to 20. 

We achieve this by finding the least common multiple (LCM) of the numbers 1 through 20. The LCM of two numbers can be computed using the formula:

$$
\text{LCM}(a, b) = \frac{a \times b}{\text{GCD}(a, b)}
$$

where GCD is the greatest common divisor. We iteratively apply this formula across the range of numbers.


In [171]:
import math

def lcm(a, b):
    """Return the least common multiple of a and b."""
    return a * b // math.gcd(a, b)

# Initialize result to 1
result = 1

# Iteratively compute LCM for numbers 1 through 20
for i in range(1, 21):
    result = lcm(result, i)

print("The smallest positive number evenly divisible by all numbers from 1 to 20 is:", result)


The smallest positive number evenly divisible by all numbers from 1 to 20 is: 232792560


## DeepSeek-R1 (81 seconds)

Logic:
To find the smallest number divisible by all integers from 1 to 20, compute the Least Common Multiple (LCM) of these numbers. The LCM is derived by multiplying the highest powers of all prime factors within the range. For example:

Prime 2 appears as 2 4 = 16 2 4 =16 (since 2 5 = 32 > 20 2 5 =32>20).

Prime 3 appears as 3 2 = 9 3 2 =9 (since 3 3 = 27 > 20 3 3 =27>20).

Primes like 5, 7, etc., only need their first power (since their squares exceed 20).

Multiply all these prime powers together to get the LCM.

In [172]:
def smallest_divisible(n):
    primes = []
    for num in range(2, n + 1):
        if all(num % i != 0 for i in range(2, int(num**0.5) + 1)):
            primes.append(num)
    
    result = 1
    for p in primes:
        max_exp = 1
        while p ** (max_exp + 1) <= n:
            max_exp += 1
        result *= p ** max_exp
    return result

print(smallest_divisible(20))  # Output: 232792560

232792560
